ceph.spec.in

```diff
@@ -702,6 +714,7 @@ Summary:	RADOS block device client library
 %if 0%{?suse_version}
 Group:		System/Libraries
 %endif
+Requires:	librados2-compat = %{_epoch_prefix}%{version}-%{release}
```

Comment:
Nit: I don't think we need to require librados2-compat (librbd1 won't actually use it). We just want to ensure that any 3rd party applications that require librados2 pull in the -compat library.

Reply:
@dillaman yes and no. my concern is that 

1. user upgrades from previous pre-nautilus librbd1 to nautilus librbd1
1. and boom!

by enforcing this dependency, we can ensure this does not happen.

> We just want to ensure that any 3rd party applications that require librados2 pull in the -compat library.

i am not sure how we can ensure this without this change. we cannot force our user to uninstall librados2 and install librados2-compat for using librbd1. if you don't like this dependency, we must add something like 

- "Conflicts: librados2" for librbd1 in d/control and
- "Obsoletes: librados2" for librbd1 in ceph.spec.in.

which forces our user to recompile the librados application, and practically defeats the purpose of librados2-compat.

Reply:
The issue w/ this forced dependency is that now librbd1 will always need to be tied to librados2-compat (and we can never move forward w/ removing it). Can't we add "Obsoletes: librados2" to librados3 -- or will this prevent the installation of "librados2-compat"?

Reply:
... I guess it depends if QEMU/libvirt/etc are tied to "Requires: librados.so.2" or "Requires: librados2" (I think it's the former)

Reply:
... a la: http://rpmfind.net/linux/RPM/fedora/updates/29/x86_64/Packages/q/qemu-block-rbd-3.0.0-3.fc29.x86_64.html

Reply:
> The issue w/ this forced dependency is that now librbd1 will always need to be tied to librados2-compat (and we can never move forward w/ removing it).

i see.

> Can't we add "Obsoletes: librados2" to librados3 -- or will this prevent the installation of "librados2-compat"?

not really. i think it's a good idea. in the worst case, we can confine the dependency like `Obsoletes: librados2 (< %{the_first_nautilus_release})` .

> Requires: librados.so.2

rpm allows and even encourage such kind of dependency. but dpkg does not.

i could try to add the "Obsoletes: librados2" to librados3. but it would be controversial to claim librados3 conflicts with librados2. because they can/should co-exist in the same system without interfering with each other. as we are used to the world where libraries with different soversions can/should happily live together.

i think the another approach is to give up the attempt to enable them to co-exist -- to stop versioning libceph-common, so that they will conflict because of libceph-common. the reason i versioned it was try to enable user to install librados2 **and** librados3 in the same system.

the nice thing of un-versioned libceph-common is that librbd1 will depend on librados3, which packages libceph-common.so. and this `libceph-common.so` will conflict with the one comes with the old librados2. so problem solved by breaking the librados{2,3} co-exists use case i imagined.

i am not sure which one is worse:

- librados3 conflicts with librados2 because the spec and d/control claim so. even they are not necessarily to conflicts with each other.
- librados3 conflicts with librados2 because the developer fails/chooses to not version an internal library.

Reply:
@dillaman another question is, without the dependency from librbd of librados2-compat, who will pull this package into the system? will it be librados3? does this imply that librados3 will need to be tied with librados2-compat for non-obvious reason that existing librbd1 client will be likely using new librbd1+librados3 along with librados2?

what we have when trying to install librbd1 which depends on librados2-compat:

```
2019-01-17T03:36:44.938 INFO:teuthology.orchestra.run.mira084.stdout:--> Processing Dependency: librados2-compat = 2:14.0.1-2643.g900a8d5.el7 for package: 2:librbd1-14.0.1-2643.g900a8d5.el7.x86_64
...
2019-01-17T03:36:44.973 INFO:teuthology.orchestra.run.mira084.stdout:--> Processing Dependency: librados.so.2()(64bit) for package: 10:qemu-kvm-1.5.3-160.el7.x86_64
2019-01-17T03:36:44.979 INFO:teuthology.orchestra.run.mira084.stdout:--> Processing Dependency: librados.so.2()(64bit) for package: 10:qemu-kvm-common-1.5.3-160.el7.x86_64
2019-01-17T03:36:44.984 INFO:teuthology.orchestra.run.mira084.stdout:--> Processing Dependency: librados.so.2()(64bit) for package: 10:qemu-img-1.5.3-160.el7.x86_64
2019-01-17T03:36:44.989 INFO:teuthology.orchestra.run.mira084.stdout:---> Package librados2-compat.x86_64 2:14.0.1-2643.g900a8d5.el7 will be obsoleting
2019-01-17T03:36:44.989 INFO:teuthology.orchestra.run.mira084.stdout:---> Package libradosstriper1.x86_64 2:14.0.1-2643.g900a8d5.el7 will be installed
...
2019-01-17T03:36:45.136 INFO:teuthology.orchestra.run.mira117.stdout:---> Package librbd1.x86_64 1:10.2.5-4.el7 will be updated
2019-01-17T03:36:45.136 INFO:teuthology.orchestra.run.mira117.stdout:---> Package librbd1.x86_64 2:14.0.1-2643.g900a8d5.el7 will be an update
...
2019-01-17T03:36:45.139 INFO:teuthology.orchestra.run.mira117.stdout:--> Processing Dependency: librados2-compat = 2:14.0.1-2643.g900a8d5.el7 for package: 2:librbd1-14.0.1-2643.g900a8d5.el7.x86_64
...
2019-01-17T03:36:45.168 INFO:teuthology.orchestra.run.mira117.stdout:---> Package librados2.x86_64 1:10.2.5-4.el7 will be obsoleted
2019-01-17T03:36:45.173 INFO:teuthology.orchestra.run.mira117.stdout:--> Processing Dependency: librados.so.2()(64bit) for package: 10:qemu-kvm-1.5.3-160.el7.x86_64
2019-01-17T03:36:45.180 INFO:teuthology.orchestra.run.mira117.stdout:--> Processing Dependency: librados.so.2()(64bit) for package: 10:qemu-kvm-common-1.5.3-160.el7.x86_64
2019-01-17T03:36:45.184 INFO:teuthology.orchestra.run.mira117.stdout:--> Processing Dependency: librados.so.2()(64bit) for package: 10:qemu-img-1.5.3-160.el7.x86_64
2019-01-17T03:36:45.189 INFO:teuthology.orchestra.run.mira117.stdout:---> Package librados2-compat.x86_64 2:14.0.1-2643.g900a8d5.el7 will be obsoleting
2019-01-17T03:36:45.190 INFO:teuthology.orchestra.run.mira117.stdout:---> Package libradosstriper1.x86_64 2:14.0.1-2643.g900a8d5.el7 will be installed
...
2019-01-17T03:36:46.398 INFO:teuthology.orchestra.run.mira084.stdout:--> Finished Dependency Resolution
2019-01-17T03:36:46.623 INFO:teuthology.orchestra.run.mira084.stdout:Error: Package: 10:qemu-kvm-common-1.5.3-160.el7.x86_64 (@rhel-7-server-rpms)
2019-01-17T03:36:46.624 INFO:teuthology.orchestra.run.mira084.stdout:           Requires: librados.so.2()(64bit)
2019-01-17T03:36:46.624 INFO:teuthology.orchestra.run.mira084.stdout:           Removing: 1:librados2-10.2.5-4.el7.x86_64 (@rhel-7-server-rpms)
2019-01-17T03:36:46.624 INFO:teuthology.orchestra.run.mira084.stdout:               librados.so.2()(64bit)
2019-01-17T03:36:46.624 INFO:teuthology.orchestra.run.mira084.stdout:           Obsoleted By: 2:librados2-compat-14.0.1-2643.g900a8d5.el7.x86_64 (ceph)
2019-01-17T03:36:46.624 INFO:teuthology.orchestra.run.mira084.stdout:               Not found
2019-01-17T03:36:46.624 INFO:teuthology.orchestra.run.mira084.stdout:           Available: 1:librados2-0.80.7-2.el7.x86_64 (rhel-7-server-rpms)
2019-01-17T03:36:46.625 INFO:teuthology.orchestra.run.mira084.stdout:               librados.so.2()(64bit)
2019-01-17T03:36:46.625 INFO:teuthology.orchestra.run.mira084.stdout:           Available: 1:librados2-0.80.7-3.el7.x86_64 (rhel-7-server-rpms)
2019-01-17T03:36:46.625 INFO:teuthology.orchestra.run.mira084.stdout:               librados.so.2()(64bit)
2019-01-17T03:36:46.625 INFO:teuthology.orchestra.run.mira084.stdout:           Available: 1:librados2-0.94.5-1.el7.x86_64 (rhel-7-server-rpms)
2019-01-17T03:36:46.625 INFO:teuthology.orchestra.run.mira084.stdout:               librados.so.2()(64bit)
2019-01-17T03:36:46.625 INFO:teuthology.orchestra.run.mira084.stdout:           Available: 1:librados2-0.94.5-2.el7.x86_64 (rhel-7-server-rpms)
2019-01-17T03:36:46.626 INFO:teuthology.orchestra.run.mira084.stdout:               librados.so.2()(64bit)
```

but i inspected the rpm package manually:
```
$ rpm2cpio ./librados2-compat-14.0.1-2643.g900a8d5.el7.x86_64.rpm | cpio -idmv
./usr/lib64/librados.so.2
1 block
$ ls -l usr/lib64/librados.so.2
lrwxrwxrwx 1 kefu kefu 13 Jan 17 12:29 usr/lib64/librados.so.2 -> librados.so.3
```

will try to add
```
Provides:   librados2:%{_libdir}/librados.so.2
```
instead.

Reply:
```
Provides:   librados2:%{_libdir}/librados.so.2
```
does not help
```
2019-01-17T11:46:10.738 INFO:teuthology.orchestra.run.mira026.stdout:Error: Package: 10:qemu-kvm-common-1.5.3-160.el7.x86_64 (@rhel-7-server-rpms)
2019-01-17T11:46:10.738 INFO:teuthology.orchestra.run.mira026.stdout:           Requires: librados.so.2()(64bit)
2019-01-17T11:46:10.738 INFO:teuthology.orchestra.run.mira026.stdout:           Removing: 1:librados2-10.2.5-4.el7.x86_64 (@rhel-7-server-rpms)
2019-01-17T11:46:10.738 INFO:teuthology.orchestra.run.mira026.stdout:               librados.so.2()(64bit)
2019-01-17T11:46:10.739 INFO:teuthology.orchestra.run.mira026.stdout:           Obsoleted By: 2:librados2-compat-14.0.1-2623.g4734aba.el7.x86_64 (ceph)
2019-01-17T11:46:10.739 INFO:teuthology.orchestra.run.mira026.stdout:               Not found
2019-01-17T11:46:10.739 INFO:teuthology.orchestra.run.mira026.stdout:           Available: 1:librados2-0.80.7-2.el7.x86_64 (rhel-7-server-rpms)
2019-01-17T11:46:10.739 INFO:teuthology.orchestra.run.mira026.stdout:               librados.so.2()(64bit)
2019-01-17T11:46:10.739 INFO:teuthology.orchestra.run.mira026.stdout:           Available: 1:librados2-0.80.7-3.el7.x86_64 (rhel-7-server-rpms)
2019-01-17T11:46:10.739 INFO:teuthology.orchestra.run.mira026.stdout:               librados.so.2()(64bit)
2019-01-17T11:46:10.739 INFO:teuthology.orchestra.run.mira026.stdout:           Available: 1:librados2-0.94.5-1.el7.x86_64 (rhel-7-server-rpms)
2019-01-17T11:46:10.739 INFO:teuthology.orchestra.run.mira026.stdout:               librados.so.2()(64bit)
2019-01-17T11:46:10.740 INFO:teuthology.orchestra.run.mira026.stdout:           Available: 1:librados2-0.94.5-2.el7.x86_64 (rhel-7-server-rpms)
2019-01-17T11:46:10.740 INFO:teuthology.orchestra.run.mira026.stdout:               librados.so.2()(64bit)
```

http://qa-proxy.ceph.com/teuthology/kchai-2019-01-17_10:16:30-rados-wip-librados2-compat-distro-basic-mira/3474544/teuthology.log

Reply:
@ktdreyer @b-ranto could you shed some light on this? is there any way we can offer "librados.so.2" in the librados2-compat package? i think "librados.so.3" is also a symlink to "librados.so.3.0.0". so technically, it should be fine to have a "librados.so.2" pointing to "librados.so.3". and i also tested locally, it works fine with a qemu-img executable.

Reply:
It might be that rpmbuild is pulling the SO version directly from the library and ignoring the link. If that’s the case, we would need to link a real .so.2

Reply:
Hmm, I am getting a bit confused, here. Is librados3 fully backwards compatible with librados2 (i.e. did we only add some new symbols?)? If so then why did we change the library version? AFAIK, it could still continue to be the old librados2 just fine.

If it is not fully backwards compatible then we should not link the libraries, it would probably cause a lot of confusion at some point in the future.

If there are backwards incompatible changes and we still want the old library then we should continue to build it (and likely also package it as `compat-librados2`).

All in all, what issue are we trying to solve, here?

Reply:
@b-ranto The C++ API has been backward incompatible in librados2 so librados3 was developed to only include a C API (and the C++ API was moved to a new shared library). The issue is that numerous 3rd party packages depend on librados.so.2 so by removing librados.so.2 from the Nautilus release, things like QEMU and libvirt will be automatically removed when Ceph is upgraded.

Reply:
@dillaman In that case, we should probably stop obsoleting librados2 in Nautilus. QEMU/libvirt could continue to use librados2 provided by distro repositories (that it was also built against) then, I suppose.

Reply:
@b-ranto The problem is that you cannot have side-by-side librados2 from an older release being used with a librbd1 from a new release since librbd1 depends on the Ceph cluster connection establiblished by librados2. So we need a librados2 that is using the Nautilus code.

Reply:
@b-ranto No, it derived from this mailing list thread [1]

[1] https://marc.info/?l=ceph-devel&m=154030910118826&w=2

Reply:
@dillaman is bumping up the soversion of librbd an option? so we don't have the issue anymore. the existing QEMU and libvirt can continue linking against librbd1, and new QEMU and libvirt can opt to link against librbd2 (and librados3).

Reply:
@tchaikov I don't want to do that since then librbd would never get updated on non-bleeding edge OSs.

Reply:
@tchaikov @dillaman I was thinking along the same lines, here. +1 for me for bumping librbd as well. It would also make sense given that new librbd is not backwards-compatible with old librados.

Reply:
@b-ranto That would mean RHEL 9 at the earliest -- terrible

Reply:
@dillaman What would be so bad about librbd1 not being updated? It would be used only by the apps in base RHEL that were built against it anyway, right? Anything else could link against the new updated librbd2.

Reply:
@b-ranto It means that QEMU/libvirt will never see any feature updates post-Mimic (basically, why have we been wasting time developing things if they can never be used)

Reply:
@dillaman I guess QEMU/libvirt could be updated to look for the new library if it wants to use any new features. I suppose it would need to be updated to actually use those features anyway, right?

Also, given that ceph can be deployed as a container, qemu/libvirt would not see these features om that scenario anyway.

I really dislike breaking RHEL ecosystem because of this -- the new library is not (fully) compatible with the old one. There might be (C++-based) 3rd party solutions that link against librados2 (I mean, there very likely already are) and this would break them.

Reply:
@b-ranto It's not new API services that QEMU doesn't yet use -- it's internal changes that would prevent QEMU from even opening RBD images using new features. Like I said, it's basically saying all RBD improvements since Mimic are now worthless.

Reply:
This is definitely being tricky. If we are really going this way then personally, I would prefer it if we just shipped the `librados.so.2` symlink in the `librados3` library and then we would set `Obsoletes:` and `Provides:` for the `librados2` package to the `librados3` package. The `librados2()` rpm dependencies should then get resolved automatically.

That looks like the simplest way to achieve this for me. We should also make sure we are not breaking anything else that might be compiled against librados2 in RHEL 7 (or RHEL 8). Adding yet another package just for the symlink seems like a bit of an overkill to me.